### PR TITLE
feat!: use w3up-client in keyring

### PR DESCRIPTION
**Pull request by travis**
per https://github.com/web3-storage/w3ui/issues/295 rework the keyring to use w3up-client rather than the lower level access library

I'd like to refactor these APIs soon, but for now I'm keeping them unchanged to minimize downstream changes

I am releasing this as a breaking change - I don't think anything will break, but would prefer to let downstream users opt in to this

### CODE REVIEW

File: packages/keyring-core/test/agent.spec.ts [hunk 1]
```diff
@@ -1,23 +1,18 @@
...
+  const client = await createClient()
+  const space = await client.createSpace('test')
   expect(space).toBeTruthy()
-  expect(space.did.startsWith('did:key:')).toBe(true)
-})
-
-test('registerSpace fails if no current space is set', async () => {
-  const agent = await createAgent()
-  await expect(agent.registerSpace('foo@bar.net')).rejects.toThrowError()
+  expect(space.did().startsWith('did:key:')).toBe(true)
```
Comment (alanshaw) [inline packages/keyring-core/test/agent.spec.ts:17]: Rename this file?

File: packages/react-keyring/src/Authenticator.tsx [hunk 2]
```diff
@@ -85,6 +85,11 @@ export const AgentLoader = ({
   return children
 }
 
+/**
+ * @deprecated use ClientLoader
```
Comment (alanshaw) [inline packages/react-keyring/src/Authenticator.tsx:89]: This is a breaking change anyways, just remove it!

File: packages/react-keyring/src/providers/Keyring.tsx [hunk 3]
```diff
@@ -81,37 +78,36 @@ export function KeyringProvider ({
...
 
   const authorize = async (email: `${string}@${string}`): Promise<void> => {
-    const agent = await getAgent()
+    const c = await getClient()
     const controller = new AbortController()
     setRegisterAbortController(controller)
 
     try {
-      await accessAuthorize(agent, email, { signal: controller.signal })
+      await c.authorize(email, { signal: controller.signal })
```
Comment (alanshaw) [inline packages/react-keyring/src/providers/Keyring.tsx:105]: Did this not change to `login`?

File: packages/react-keyring/src/providers/Keyring.tsx [hunk 4]
```diff
@@ -129,62 +125,65 @@ export function KeyringProvider ({
...
   }
 
-  const registerSpace = async (email: string, opts: RegisterSpaceOpts = {}): Promise<void> => {
-    const agent = await getAgent()
-    await agent.registerSpace(email, {
-      provider: opts.provider ?? (agent.connection.id.did() as DID<'web'>)
-    })
-    setSpace(getCurrentSpaceInAgent(agent))
-    setSpaces(getSpaces(agent))
+  const registerSpace = async (email: string): Promise<void> => {
```
Comment (alanshaw) [inline packages/react-keyring/src/providers/Keyring.tsx:136]: Maybe this should be renamed to `provisionSpace`?

File: packages/react-keyring/src/providers/Keyring.tsx [hunk 5]
```diff
@@ -129,62 +125,65 @@ export function KeyringProvider ({
...
+    const c = await getClient()
+    const account = useAccount(c, { email })
+    const space = c.currentSpace()
+    if (account && space) {
+      await account.provision(space.did() as DIDKey)
+      setSpace(c.currentSpace())
+      setSpaces(c.spaces())
+    }
```
Comment (alanshaw) [inline packages/react-keyring/src/providers/Keyring.tsx:144]: Need to throw if missing account or space.

File: packages/react-keyring/src/providers/Keyring.tsx [hunk 6]
```diff
@@ -129,62 +125,65 @@ export function KeyringProvider ({
...
   const setCurrentSpace = async (did: DID): Promise<void> => {
-    const agent = await getAgent()
-    await agent.setCurrentSpace(did as DID<'key'>)
-    setSpace(getCurrentSpaceInAgent(agent))
+    const c = await getClient()
+    await c.setCurrentSpace(did as DID<'key'>)
+    setSpace(c.currentSpace())
   }
 
   const loadAgent = async (): Promise<void> => {
```
Comment (alanshaw) [inline packages/react-keyring/src/providers/Keyring.tsx:153]: Rename `loadClient`?

File: packages/react-keyring/src/providers/Keyring.tsx [hunk 7]
```diff
@@ -129,62 +125,65 @@ export function KeyringProvider ({
...
-    if (agent != null) return
-    await getAgent()
+    if (client != null) return
+    await getClient()
   }
 
   const unloadAgent = async (): Promise<void> => {
```
Comment (alanshaw) [inline packages/react-keyring/src/providers/Keyring.tsx:158]: Rename `unloadClient`

File: packages/react-keyring/src/providers/Keyring.tsx [hunk 8]
```diff
@@ -129,62 +125,65 @@ export function KeyringProvider ({
...
     setSpace(undefined)
     setSpaces([])
-    setIssuer(undefined)
-    setAgent(undefined)
+    setClient(undefined)
     setAccount(undefined)
   }
 
   const resetAgent = async (): Promise<void> => {
```
Comment (alanshaw) [inline packages/react-keyring/src/providers/Keyring.tsx:165]: Rename `resetClient`?

**Review by alanshaw — CHANGES_REQUESTED** [review]
Looks good so far!

**Comment by travis** [comment]
@alanshaw re: a lot of this - I'd like to avoid changing the API here so we don't have to go update all the examples. I hate the idea of trying to design the next version of this API under pressure of launching, so I propose we make these changes without any changes to the interfaces these modules export and then fast follow with a breaking change that mostly just renames various functions (but also potentially reworks/deletes/adds functions that makes sense)

what do you think?

File: packages/react-keyring/src/providers/Keyring.tsx (hunk 3 above)
Comment (travis) [reply packages/react-keyring/src/providers/Keyring.tsx:109]: HM you're right - it looks like `authorize` wasn't removed, but I do think this should use login - will update!

**Review by alanshaw — APPROVED** [review]

**Pull request merged**